File: Backend/models/product.js
const { DataTypes } = require("sequelize");
const { sequelize } = require("../config/db");

const Product = sequelize.define(
  "Product",
  {
    id: {
      type: DataTypes.UUID,
      defaultValue: DataTypes.UUIDV4,
      primaryKey: true,
    },
    skuCode: {
      type: DataTypes.STRING,
      allowNull: false,
      unique: true,
      trim: true,
    },
    name: {
      type: DataTypes.STRING,
      allowNull: false,
      trim: true,
    },
    category: {
      type: DataTypes.STRING,
      allowNull: false,
      trim: true,
    },
    unitOfMeasure: {
      type: DataTypes.STRING,
      allowNull: false,
      trim: true,
    },
    reorderPoint: {
      type: DataTypes.INTEGER,
      defaultValue: 0,
      validate: {
        min: 0,
      },
    },
    reorderQuantity: {
      type: DataTypes.INTEGER,
      defaultValue: 0,
      validate: {
        min: 0,
      },
    },
    stockByLocation: {
      type: DataTypes.JSONB,
      defaultValue: {},
    },
  },
  {
    timestamps: true,
  }
);

module.exports = Product;

File: Frontend/lib/features/products/models/product.dart
class Product {
  const Product({
    required this.skuCode,
    required this.name,
    required this.category,
    required this.unitOfMeasure,
    required this.reorderPoint,
    required this.reorderQuantity,
    required this.stockByLocation,
  });

  final String skuCode;
  final String name;
  final String category;
  final String unitOfMeasure;
  final int reorderPoint;
  final int reorderQuantity;
  final Map<String, int> stockByLocation;

  String get id => skuCode;

  int get totalStock {
    return stockByLocation.values.fold<int>(0, (int sum, int qty) => sum + qty);
  }

  Product copyWith({
    String? skuCode,
    String? name,
    String? category,
    String? unitOfMeasure,
    int? reorderPoint,
    int? reorderQuantity,
    Map<String, int>? stockByLocation,
  }) {
    return Product(
      skuCode: skuCode ?? this.skuCode,
      name: name ?? this.name,
      category: category ?? this.category,
      unitOfMeasure: unitOfMeasure ?? this.unitOfMeasure,
      reorderPoint: reorderPoint ?? this.reorderPoint,
      reorderQuantity: reorderQuantity ?? this.reorderQuantity,
      stockByLocation: stockByLocation ?? this.stockByLocation,
    );
  }
}
